draft-ietf-resumable-uploads-protocol.md

```diff
@@ -177,96 +188,70 @@ If the request completes successfully, the server MUST acknowledge it by respond
-
-The request MUST use the `HEAD` method and include the `Upload-Token` header. The request MUST NOT include the `Upload-Offset` header or the `Upload-Incomplete` header.
-
-If the server has resources allocated for this token, it MUST send back a `204 (No Content)` response with a header `Upload-Offset` which indicates the resumption offset for the client. If the server has multiple discontiguous chunks of the same file, the offset MUST be the end of the first chunk.
-
-The response SHOULD include `Cache-Control: no-store` header to prevent HTTP caching.
+:status: 104
 
-ABNF of `Upload-Offset` is
-
-```
-Upload-Offset = sf-integer
+:status: 201
 ```
 
-If the server has no record of this token, it MUST respond with `404 (Not Found)` status code.
-
 ```
-:method: HEAD
+:method: POST
 :scheme: https
 :authority: example.com
 :path: /upload
 upload-token: :SGVs…SGU=:
+upload-offset: 0
+upload-incomplete: ?1
+[partial file content]
 
-:status: 204
-upload-offset: 100
-cache-control: no-store
+:status: 201
+upload-incomplete: ?1
 ```
 
-The client MAY automatically start uploading from the beginning using Upload Creation Procedure if `404 (Not Found)` status code is received. The client SHOULD NOT automatically retry if a status code other than 204 and 404 is received.
+The client MAY automatically attempt upload resumption when the connection is terminated unexpectedly, or if a server error status code between 500 and 599 (inclusive) is received. The client SHOULD NOT automatically retry if a client error status code between 400 and 499 (inclusive) is received.
 
-## Upload Appending Procedure
+### Feature Detection
 
-Upload Appending Procedure can be used for either resuming an existing upload, or starting a new upload. The client MUST NOT initiate this procedure without the knowledge of server support [Service Discovery].
+If the client has no knowledge of whether the server supports resumable upload, the Upload Transfer Procedure MAY be used with some additional constraints. In particular, the `Upload-Offset` header and the `Upload-Incomplete` header MUST NOT be sent in the request if the server support is unclear. This allows the upload to function as if it is a regular upload.
 
-The request MUST use the `PATCH` method and include both the `Upload-Token` and the `Upload-Offset` header. The `Upload-Incomplete` header MAY be used if the request body is not the complete file. ABNF of `Upload-Incomplete` is
+If the server detects the Upload Transfer Procedure with neither the `Upload-Offset` header nor the `Upload-Incomplete` header, and it supports resumable upload, an informational response with `104 (Upload Resumption Supported)` status MAY be sent to the client while the request body is being uploaded.
```

Comment:
I still don't think this 104 response is a good idea as support for this is currently non-existant. Do we know how existing proxies, clients and browsers would handle cases when they suddenly encounter a 104? Do they just ignore it or error out? That must be clarified before we include this section in the spec.

Reply:
Clients and browsers will only receive 104 if their request includes `Upload-Token` header. This means they are aware of the resumable upload standard.

Proxy I'm not familiar with. It is possible that some of them drop 1xx responses instead of forwarding them, but that still allows the upload to proceed as a regular upload.

104 feature detection is completely optional. You can disable it by sending `Upload-Offset: 0` during upload creation. I went through all the options, and I think this is an upgrade from SETTINGS where HTTP stack has to be modified. And it is also the only option that avoids an extra round trip.